patch felixconfig with env vars

ozdanborne opened the pull request:
## Description

This PR notices `FELIX_*` env vars in calico-node and "moves" them into FelixConfig using PATCH directives.

There's two reasons this is extra tricky:
- Felix Env vars are in all caps lock, so it's unclear which FelixConfig field to set. To resolve this, we use Go Reflection to find the struct field which, when downcased, matches the env var name.
- Feli Env vars are strings. We need to convert them to string / number / bool, the three primatives used in JSON. To accomplish this, again we use reflection to check which type the FelixConfig field is, and then use various methods to convert the data correctly.

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

ozdanborne commented on pkg/controller/migration/convert/felix_vars.go:
```diff
@@ -0,0 +1,143 @@
+		if strings.ToLower(key) == strings.ToLower(field.Name) {
+			fieldName := strings.Split(field.Tag.Get("json"), ",")[0]
+
+			v, err := convert(field.Type, value, val)
```
I think passing both `fc.Type().Field(ii).Type` and `fc.Field(ii)` is redundant since you can go from `reflect.Value` to `reflect.Type`. 

But... I keep hitting panics when I do that. So I've gotta spend more time and figure out how.

ozdanborne commented on pkg/controller/migration/convert/convert_test.go:
```diff
@@ -45,14 +45,14 @@ var _ = Describe("Parser", func() {
 				Namespace: "kube-system",
 			},
-		}, emptyKubeControllerSpec(), pool)
+		}, emptyKubeControllerSpec(), pool, emptyFelixConfig())
```
a `replace` patch will fail if the resource does not exist. I can add code to handle that by creating the default felixconfig if it does not exist. But I'm surprised to hear you don't think one will always exist. It seemed like calico-node startup should always have created one: https://github.com/projectcalico/node/blob/master/pkg/startup/startup.go#L1106

ozdanborne commented on pkg/controller/migration/convert/felix_vars.go:
```diff
@@ -0,0 +1,192 @@
+			return &str, nil
+		}
+		return str, nil
+	case *crdv1.IptablesBackend:
+		v := crdv1.IptablesBackend(str)
+		return &v, nil
+	case *crdv1.AWSSrcDstCheckOption:
+		v := crdv1.AWSSrcDstCheckOption(str)
+		return &v, nil
```
I'm not sure if this is OK or if I need to run a switch statement on the value and return the actual 'enum' it corresponds to.

tmjd commented on pkg/controller/migration/convert/felix_vars_test.go:
```diff
@@ -0,0 +1,156 @@
+		}))
+	})
+
+	Context("", func() {
```
Is there some kind of Context we can have here?

tmjd commented on pkg/controller/migration/convert/felix_vars.go:
```diff
@@ -0,0 +1,143 @@
+			}
+
+			return patch{
+				Op:    "replace",
```
I'd appreciate at least one real k8s-apiserver for this code to make sure it doesn't behave different than the Fake does for patches. My understanding is that the patch is submitted to the API server, that's why I'm concerned there might be different behavior.

ozdanborne replied:
Yep, confirmed using kubectl:

```
(⎈ |kind-turk-bz-7vui:default)❯ k get felixconfiguration default -o yaml
apiVersion: crd.projectcalico.org/v1
kind: FelixConfiguration
metadata:
  name: default
spec:
  logSeverityScreen: warn
  reportingInterval: 30s
```

Replacing `iptablesBackend` without issue even tohugh it's not set yet :+1: 

```
(⎈ |kind-turk-bz-7vui:default)❯ kubectl patch felixconfigurations.crd.projectcalico.org default --type json -p '[{"op":"replace","path":"/spec/iptablesBackend","value":"Legacy"}]' 
felixconfiguration.crd.projectcalico.org/default patched
```

ozdanborne commented on pkg/controller/migration/convert/felix_vars.go:
```diff
@@ -0,0 +1,192 @@
+			if len(vals) != 2 {
+				return nil, fmt.Errorf("could not convert protoport: must be of form <proto>:<port>")
+			}
+			port, err := strconv.Atoi(vals[1])
```
Nice, I didn't know there was `ParseUint`

ozdanborne commented:
While addressing feedback, I remembered that I had commented out some `numorstring` references in FelixConfig. To resolve this, I copied the bits of the numorstring library that were needed for parsing values. This way, we don't have to depend on libcalico-go.

tmjd commented on pkg/controller/migration/convert/convert.go:
```diff
@@ -140,6 +140,10 @@ func Convert(ctx context.Context, client client.Client, install *operatorv1.Inst
 		}
 	}
 
+	if err := handleFelixVars(comps); err != nil {
```
I think it would be good to have a comment here explaining why this is separate from the other handlers. Maybe something like `Handle the FelixVars as the last handler because we take values and directly load them to the datastore so we want to know that we have migrated all the other config successfully`
